### ISSUE TITLE
Add COMBINE primitive as alternative to REJOIN (as opposed to changing REJOIN)

### ISSUE THREAD

**Issue by rebolbot** [rebol/rebol-issues#2142]
_Submitted by:_ **fork**

Frustrations with the "ugliness" of REJOIN's name and behavior led me to make #2079 and #2100 (to try and take other names or handle NONE differently).  After the defense of JOIN being a binary operator which is useful and the disruption to legacy (a case I see now), I've dismissed both tickets in favor of a proposal to pursue this line of thinking with an entirely new primitive:

``` rebol
>> combine [{abc} if false {def} {ghi}]
== {abcghi} ;-- contrast this with REJOIN's {abcnoneghi}
```

``` rebol
>> combine [[a b c] if/only false [d e f] [g h i]]
== [a b c g h i] ;-- contrast with REJOIN's [a b c none [g h i]]
```

The general rule would be that combine would throw out nones.  Merging unequal types would likely otherwise follow rejoin's logic.  The type would match the first non-none.  If all the elements were none (or no elements) the result would be a none.

``` rebol
>> combine []
== none
```

``` rebol
>> a: b: c: none
>> combine [a b c]
== none
```

There would be a /WITH refinement for content to be combined with an intermediary element, effectively placing an element in-between as if it had been there.  Same logic would apply:

``` rebol
>> combine/with [{abc} none {def} {ghi}] {,}
== {abc,def,ghi}
```

``` rebol
>> combine/with [[a b c] none [d e f] [g h i]] [foo bar]
== [a b c foo bar d e f foo bar g h i]
```

The rules of nested blocks if producing a string series would be to run the combine continuously until things were flattened (as opposed to FORM-ing):

``` rebol
>> stuff: [{baz} {bar}]
>> combine [{foo} stuff]
== {foobazbar} ;-- as opposed to {foobaz bar} or {foo[{baz} {bar}]} or whatever
```

Getting such behavior on blocks may not be what is desired, so being able to suppress it with a /SHALLOW, or make it default to not doing it for block series and adding a /DEEP refinement would be better.

/ONLY might control whether things like paths would be collapsed in with blocks, if that were the default:

``` rebol
>> combine [a/b/c [d e f] (g h i)]
== a/b/c/d/e/f/g/h/i
```

``` rebol
>> combine/only [a/b/c [d e f] (g h i)]
== a/b/c/[d e f]/(g h i)
```

The idea of merging the series by default is in that area of things that are kind of ugly, but consistent.

An /ALL refinement or similar could suggest nones not be excluded, which would help people who were looking for COMBINE's behavior who didn't want to switch to REJOIN when they needed a none.

``` rebol
>> combine/all [[a b c] none [d e f]]
== [a b c none d e f]
```

---

Credit for the idea of making a new primitive instead of struggling with REJOIN from @earl.  Name COMBINE from @Gregg.  Maybe FLATTEN, which makes sense if the /DEEP behavior is the default for blocks.  (Other suggestions of CONCAT, RECAT, or CATENATE don't seem as nice to me.)  /WITH suggested by @rgchris.

<sup>**CC - Data** [ Version: r3 master Type: Wish Platform: All Category: Native Reproduce: Always Fixed-in:none ]</sup>

**Comment by maxim** [comment]
Folks, just FYI, the names in your "Submitted by" don't match up with Github usernames, so I've gotten some mentions.

**Comment by ladislav** [comment]
The contents of this article: https://github.com/ladislav/rebol.net/blob/master/wikipedia/Replacement.wiki are related.

**Comment by hostilefork** [comment]
*(Sorry you're getting mentioned, GitHub maxim.  Hopefully doesn't happen much now.)*

As for this issue, I'm going to close it (as I opened it long ago).  One major motivator for rethinking REJOIN...the rendering of "none"s, has been systemically cured with a new [NULL](https://github.com/metaeducation/ren-c/commit/86ca061dfbe0c58e63b13a510f302bc680a3c265) policy.  There are now [SPACED, UNSPACED, and DELIMIT](https://trello.com/c/QOAJa1n1) functions, which have pretty decent names for common cases.

Also, I noticed that I kept mixing up COMBINE and COMPOSE when I was using them both, just typing the wrong one or getting confused.  I'd be less likely to be confused nowadays, but if I could be confused so could someone else.

Long story short: a lot of the questions that were open issues in the COMBINE era are resolved, so it's probably best to bear Ladislav's link in mind for anyone needing more complex behaviors, and I will close this.
